src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -139,66 +132,58 @@ private async Task GetCpuInfoAsync()
+                List<float[]> samples = new();
+                int samplesCount = 10;
 
-                using var timeoutTokenSource = new CancellationTokenSource();
-                timeoutTokenSource.CancelAfter(TimeSpan.FromMilliseconds(METRICS_UPDATE_INTERVAL));
-
-                using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
-                    _context.CancellationToken,
-                    timeoutTokenSource.Token);
-
-                processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
+                // /proc/stat updates linearly in real time and shows CPU time counters during the whole system uptime
+                // so we need to collect multiple samples to calculate CPU usage
+                for (int i = 0; i < samplesCount + 1; i++)
                 {
-                    var processInvokerOutput = message.Data;
-
-                    var cpuInfoNice = int.Parse(processInvokerOutput.Split(' ', (char)StringSplitOptions.RemoveEmptyEntries)[2]);
-                    var cpuInfoIdle = int.Parse(processInvokerOutput.Split(' ', (char)StringSplitOptions.RemoveEmptyEntries)[4]);
-                    var cpuInfoIOWait = int.Parse(processInvokerOutput.Split(' ', (char)StringSplitOptions.RemoveEmptyEntries)[5]);
-
-                    lock (_cpuInfoLock)
+                    string[] strings = await File.ReadAllLinesAsync("/proc/stat", cancellationToken);
+                    if (cancellationToken.IsCancellationRequested)
                     {
-                        _cpuInfo.Updated = DateTime.Now;
-                        _cpuInfo.Usage = (double)(cpuInfoNice + cpuInfoIdle) * 100 / (cpuInfoNice + cpuInfoIdle + cpuInfoIOWait);
+                        return;
                     }
-                };
 
-                processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs message)
+                    samples.Add(strings[0]
+                            .Split(' ', StringSplitOptions.RemoveEmptyEntries)
+                            .Skip(1)
+                            .Select(float.Parse)
+                            .ToArray());
+
+                    await Task.Delay(100, cancellationToken);
+                }
+
+                // The CPU time counters in the /proc/stat are:
+                // user, nice, system, idle, iowait, irq, softirq, steal, guest, guest_nice
+                //
+                // We need to get deltas for idle and total CPU time using the gathered samples
+                // and calculate the average to provide the CPU utilization in the moment
+                double cpuUsage = 0.0;
+                for (int i = 1; i < samplesCount + 1; i++)
                 {
-                    Trace.Error($"Error on receiving CPU info: {message.Data}");
-                };
+                    double idle = samples[i][3] - samples[i - 1][3];
```

Comment:
One thing to check: there may be risk of index to of range here if the array is not in the expected shape.  it's better to check length of the array to ensure the indexes are in range to avoid the risk.  Alternative is catch IndexOutOfRange exception.

Reply:
The index out of range issue should not happen here because of two reasons - the first index is the index of the collected sample, and we always gather `samplesCount` + 1 before executing the calculation loop. The only case when we can get lesser number of samples if the previous loop will be canceled due to timeout but in this case the whole method will be canceled, and we do not enter in this loop. The second index is responsible for the `/proc/stat` field, and the output from there may be not available in two cases: if the `/proc/stat` is unavailable which signalize that the OS is broken, and if the output from CPU counter has changed, which may happen extremely rarely only with major kernel updates.

Reply:
There is an exception wrapper on calls to GetCpuInfoAsync, so I'm okay with this.